x/cronos/keeper/precompiles/ica.go

```diff
@@ -88,16 +95,35 @@ func (ic *IcaContract) Address() common.Address {
 }
 
 // RequiredGas calculates the contract gas use
-func (ic *IcaContract) RequiredGas(input []byte) uint64 {
-	// base cost to prevent large input size
-	baseCost := uint64(len(input)) * ic.kvGasConfig.WriteCostPerByte
+// write `max(0, len(input) * DefaultTxSizeCostPerByte + requiredGasTable[methodPrefix] - intrinsicGas)`
+// query `len(input) * ReadCostPerByte + HasCost + ReadFlat`
+func (ic *IcaContract) RequiredGas(input []byte) (gas uint64) {
 	var methodID [4]byte
 	copy(methodID[:], input[:4])
 	requiredGas, ok := icaGasRequiredByMethod[methodID]
-	if ok {
-		return requiredGas + baseCost
+	if !ok {
+		requiredGas = 0
+	}
+	// base cost to prevent large input size
+	baseCost := uint64(0)
+	method := icaMethodNamedByMethod[methodID]
+	intrinsicGas, err := core.IntrinsicGas(input, nil, false, true, true)
+	if err != nil {
+		return 0
+	}
+	if method == QueryAccountMethodName {
+		baseCost = uint64(len(input)) * ic.kvGasConfig.ReadCostPerByte
```

Comment:
@yihuang we only can calculate [Has](https://github.com/cosmos/ibc-go/blob/v7.3.1/modules/apps/27-interchain-accounts/controller/keeper/keeper.go#L212) cost based on [input](https://github.com/cosmos/ibc-go/blob/v7.3.1/modules/apps/27-interchain-accounts/controller/keeper/keeper.go#L210), not sure how to get the [Get](https://github.com/cosmos/ibc-go/blob/v7.3.1/modules/apps/27-interchain-accounts/controller/keeper/keeper.go#L216) cost since it's based on the result.